packages/core-app-api/src/apis/implementations/IdentityApi/AppIdentityProxy.ts

```diff
@@ -53,38 +53,38 @@ export class AppIdentityProxy implements IdentityApi {
   }
 
   async getProfileInfo(): Promise<ProfileInfo> {
-    if (!this.target) {
-      throw mkError('getProfileInfo');
-    }
-    return this.target.getProfileInfo();
+    return this.waitForTarget().then(target => target.getProfileInfo());
   }
 
   async getBackstageIdentity(): Promise<BackstageUserIdentity> {
-    if (!this.target) {
-      throw mkError('getBackstageIdentity');
-    }
-    return this.target.getBackstageIdentity();
+    return this.waitForTarget().then(target => target.getBackstageIdentity());
   }
 
   async getCredentials(): Promise<{ token?: string | undefined }> {
-    if (!this.target) {
-      throw mkError('getCredentials');
-    }
-    return this.target.getCredentials();
+    return this.waitForTarget().then(target => target.getCredentials());
   }
 
   async getIdToken(): Promise<string | undefined> {
-    if (!this.target) {
-      throw mkError('getIdToken');
-    }
-    return this.target.getIdToken();
+    return this.waitForTarget().then(target => target.getIdToken());
   }
 
   async signOut(): Promise<void> {
-    if (!this.target) {
-      throw mkError('signOut');
-    }
-    await this.target.signOut();
+    await this.waitForTarget().then(target => target.signOut());
     location.reload();
   }
+
+  private waitForTarget(): Promise<IdentityApi> {
+    if (this.target) {
+      return Promise.resolve(this.target);
+    }
+
+    return new Promise<IdentityApi>(resolve => {
+      const interval = setInterval(() => {
+        if (this.target) {
+          clearInterval(interval);
+          resolve(this.target);
+        }
+      }, 50);
```

Comment:
Very open to other ideas.

Reply:
Useful in cases like this: [make a deferrable](https://github.com/backstage/backstage/blob/68eddf88a02a2244cf10e64a7dd969133ddd64a1/plugins/catalog-react/src/components/UnregisterEntityDialog/useUnregisterEntityDialogState.test.tsx#L136) so you can store the promise itself, just pass that out to callers, and resolve it at will instantly.